### PR TITLE
Swagger plugins support

### LINKED ISSUE #3049: [Feature request]: Expose `plugins` settings in swagger ui config API
### Is your feature request related to a specific problem? Or an existing feature?

Please expose `plugins` settings in swagger ui configObject and render them in `index.js` similar to js `requestInterceptor` /`responseInterceptor`.

### Describe the solution you'd like

I want to be able to configure swagger ui with custom plugins (via c# API of `ConfigObject` or `SwaggerUIOptions`).
When rendered, the js code should look like 
````js
configObject.plugins=[myCustomPlugin1,myCustomPlugin2];
````

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jvmlet**
# Pull Request

## The issue or feature being addressed

Closes #3049

### CODE REVIEW

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs [hunk 1]
```diff
@@ -190,6 +190,13 @@ public class ConfigObject
         [JsonPropertyName("validatorUrl")]
         public string ValidatorUrl { get; set; } = null;
 
+
+        /// <summary>
+        /// Custom plugins function names.
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs:195]:
```suggestion
        /// Any custom plugins' function names.
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs [hunk 2]
```diff
@@ -190,6 +190,13 @@ public class ConfigObject
         [JsonPropertyName("validatorUrl")]
         public string ValidatorUrl { get; set; } = null;
 
+
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs:193]:
```suggestion
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs [hunk 3]
```diff
@@ -190,6 +190,13 @@ public class ConfigObject
...
+        /// <summary>
+        /// Custom plugins function names.
+        /// </summary>
+        [JsonPropertyName("plugins")]
+        public string[] Plugins { get; set; } = null;
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptions.cs:198]:
```suggestion
        public IList<string> Plugins { get; set; } = null;
```

File: README.md [hunk 4]
```diff
@@ -1302,13 +1302,16 @@ app.UseSwaggerUI(c =>
...
     c.ShowCommonExtensions();
+    c.Plugins = new string[]{"myCustomPlugin"}
     c.EnableValidator();
     c.SupportedSubmitMethods(SubmitMethod.Get, SubmitMethod.Head);
     c.UseRequestInterceptor("(request) => { return request; }");
     c.UseResponseInterceptor("(response) => { return response; }");
 });
 ```
 
+_NOTE: When adding custom plugins, make sure you added custom `js` file that defines the plugin function._
```
Comment (martincostello) [inline README.md:1313]:
```suggestion
> [!NOTE]  
> When adding custom plugins, make sure you add any custom `js` files that define the plugin function(s)._
```

File: README.md [hunk 5]
```diff
@@ -1302,13 +1302,16 @@ app.UseSwaggerUI(c =>
     c.MaxDisplayedTags(5);
     c.ShowExtensions();
     c.ShowCommonExtensions();
+    c.Plugins = new string[]{"myCustomPlugin"}
```
Comment (martincostello) [inline README.md:1305]:
```suggestion
    c.Plugins = ["myCustomPlugin"];
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.js [hunk 6]
```diff
@@ -57,6 +57,7 @@ window.onload = function () {
     if (interceptors.ResponseInterceptorFunction)
         configObject.responseInterceptor = parseFunction(interceptors.ResponseInterceptorFunction);
 
+    configObject.plugins = configObject.plugins ? configObject.plugins.map(eval) : [];
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/index.js:60]: Would it be better to omit setting it at all?

```suggestion
    if (configObject.plugins) {
        configObject.plugins = configObject.plugins.map(eval);
    }
```

File: test/WebSites/CustomUIConfig/Startup.cs [hunk 7]
```diff
@@ -73,14 +73,15 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
                 // Other
                 c.DocumentTitle = "CustomUIConfig";
                 c.StylesPath = "/ext/custom-stylesheet.css";
-                c.ScriptBundlePath = "/ext/custom-javascript.js";
-                c.ScriptPresetsPath = "/ext/custom-javascript.js";
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/Startup.cs:77]: Why did you remove these?

File: test/WebSites/CustomUIConfig/Startup.cs [hunk 8]
```diff
@@ -73,14 +73,15 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
                 c.InjectStylesheet("/ext/custom-stylesheet.css");
                 c.InjectJavascript("/ext/custom-javascript.js");
+                c.InjectJavascript("/ext/custom-plugin.js");
                 c.UseRequestInterceptor("(req) => { req.headers['x-my-custom-header'] = 'MyCustomValue'; return req; }");
                 c.UseResponseInterceptor("(res) => { console.log('Custom interceptor intercepted response from:', res.url); return res; }");
                 c.EnablePersistAuthorization();
 
+                c.ConfigObject.Plugins = new[] { "customPlugin" };
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/Startup.cs:83]:
```suggestion
                c.ConfigObject.Plugins = ["customPlugin"];
```

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js:1]:
```suggestion
```

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+const customPlugin = function (system) {
+    var elem = document.createElement("div");
+    elem.innerHTML =
+        "<div style=\"text-align: center; font-family: Titillium Web,sans-serif; margin: 16px;\">This text was injected via /ext/custom-plugin.js, using the SwaggerUIOptions.Plugins method.</div>";
+
+    document.body.insertBefore(elem, document.body.firstChild);
+    return {
+         
+    }
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js:10]:
```suggestion
    return {
    };
```

**Review by martincostello — COMMENTED** [review]
Please add tests.

**Comment by jvmlet** [comment]
> Please add tests.

There is no infrastructure in integration tests to evaluate the `index.html`  in browser (like `Selenium`) and I also don't see tests for request/response interceptors.

File: README.md (hunk 4 above)
Comment (jvmlet) [reply README.md:1313]: `[!NOTE]` is not rendered

**Comment by martincostello** [comment]
> I also don't see tests for request/response interceptors.

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/d7a92a121d6c3c22c3d880b75c05dda52832a819/test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs#L100-L110

**Comment by jvmlet** [comment]
> Please add tests.

There is no infrastructure in integration tests to evaluate the index.html in browser (like Selenium) and I also don't see tests for request/response interceptors.

File: test/WebSites/CustomUIConfig/Startup.cs (hunk 7 above)
Comment (martincostello) [reply test/WebSites/CustomUIConfig/Startup.cs:77]: Then you need to add a different test, because you're changing our test coverage and it's also going to break this test:

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/d7a92a121d6c3c22c3d880b75c05dda52832a819/test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs#L132-L151

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.js [hunk 11]
```diff
@@ -57,7 +57,9 @@ window.onload = function () {
     if (interceptors.ResponseInterceptorFunction)
         configObject.responseInterceptor = parseFunction(interceptors.ResponseInterceptorFunction);
 
-    configObject.plugins = configObject.plugins ? configObject.plugins.map(eval) : [];
+    if(configObject.plugins) {
+        configObject.plugins = configObject.plugins.map(eval);
+    }
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/index.js:62]:
```suggestion
    if (configObject.plugins) {
        configObject.plugins = configObject.plugins.map(eval);
    }

```

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js [hunk 12]
```diff
@@ -6,6 +6,5 @@ const customPlugin = function (system) {
 
     document.body.insertBefore(elem, document.body.firstChild);
     return {
-         
     }
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js:9]:
```suggestion
    };
```

**Comment by jvmlet** [comment]
> > I also don't see tests for request/response interceptors.
> 
> https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/d7a92a121d6c3c22c3d880b75c05dda52832a819/test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs#L100-L110

Great, I would expect to see test that actually asserts the  js function functionality , not the template. With plugins it always there ....

**Comment by martincostello** [comment]
> I would expect to see test that actually asserts the js function functionality

That would be testing swagger-ui itself, which isn't our code.

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js (hunk 12 above)
Comment (jvmlet) [reply test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js:9]: image

**Comment by jvmlet** [comment]
> > I would expect to see test that actually asserts the js function functionality
> 
> That would be testing swagger-ui itself, which isn't our code.

Then how do you suggest to test this line ? Without Selenium I can assert that the plugin was loaded ...

File: README.md [hunk 13]
```diff
@@ -1309,8 +1309,8 @@ app.UseSwaggerUI(c =>
     c.UseResponseInterceptor("(response) => { return response; }");
 });
 ```
-
-_NOTE: When adding custom plugins, make sure you added custom `js` file that defines the plugin function._
+ [!NOTE]
+ When adding custom plugins, make sure you added custom `js` file that defines the plugin function.
```
Comment (martincostello) [inline README.md:1313]: Needs the `>` characters.

```suggestion
> [!NOTE]
> When adding custom plugins, make sure you add any custom `js` files that defines the plugin function(s).
```

File: src/Swashbuckle.AspNetCore.SwaggerUI/index.js [hunk 14]
```diff
@@ -60,6 +60,7 @@ window.onload = function () {
     if(configObject.plugins) {
```
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerUI/index.js:60]:
```suggestion
    if (configObject.plugins) {
```

**Comment by martincostello** [comment]
> Then how do you suggest to test this line ? Without Selenium I can't assert that the plugin was loaded ...

Do what the tests I already directed you to do - check that the text of the configured plugins is in the rendered document.

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js (hunk 12 above)
Comment (martincostello) [reply test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin.js:9]: That's an entirely different line of code.

File: README.md [hunk 15]
```diff
@@ -1309,8 +1309,9 @@ app.UseSwaggerUI(c =>
     c.UseResponseInterceptor("(response) => { return response; }");
 });
 ```
- [!NOTE]
- When adding custom plugins, make sure you added custom `js` file that defines the plugin function.
+
+ > [!NOTE]
+ > When adding custom plugins, make sure you added custom `js` file that defines the plugin function(s).
```
Comment (martincostello) [inline README.md:1314]:
```suggestion
 > When adding custom plugins, make sure you add any custom `js` files that defines the plugin function(s).
```

Please accept the full suggestions, rather than just parts of them.

Comment (martincostello) [reply README.md:1314]: Also, please don't self-resolve the review comments - I'll resolve them when I'm happy they're dealt with as requested.

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 16]
```diff
@@ -74,6 +75,28 @@ public async Task SwaggerUIMiddleware_ReturnsInitializerScript(
...
+            var jsContent = await jsResponse.Content.ReadAsStringAsync();
+            if (string.IsNullOrEmpty(customPlugins))
+            {
+                Assert.DoesNotContain("\"plugins\"", jsContent);
+            }
+            else
+            {
+                string pluginsNames = string.Join(',', customPlugins.Split(',').Select(p => $"\"{p}\""));
+                Assert.Contains($"\"plugins\":[{pluginsNames}]", jsContent);
+            }
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:97]: Rather than have this `if` and having to compute the expected value, split this into two tests and in the second case just specify the literal string expected.

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin1.js [hunk 17]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin1.js:1]:
```suggestion
```

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin1.js [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+const customPlugin1 = function (system) {
+    var elem = document.createElement("div");
+    elem.innerHTML =
+        "<div style=\"text-align: center; font-family: Titillium Web,sans-serif; margin: 16px;\">This text was injected via /ext/custom-plugin1.js, using the SwaggerUIOptions.Plugins method.</div>";
+
+    document.body.insertBefore(elem, document.body.firstChild);
+    return {
+    };
+}
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin1.js:10]:
```suggestion
};
```

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin2.js [hunk 19]
```diff
@@ -0,0 +1,10 @@
+
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin2.js:1]:
```suggestion
```

File: test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin2.js [hunk 20]
```diff
@@ -0,0 +1,10 @@
...
+const customPlugin2 = function (system) {
+    var elem = document.createElement("div");
+    elem.innerHTML =
+        "<div style=\"text-align: center; font-family: Titillium Web,sans-serif; margin: 16px;\">This text was injected via /ext/custom-plugin2.js, using the SwaggerUIOptions.Plugins method.</div>";
+
+    document.body.insertBefore(elem, document.body.firstChild);
+    return {
+    };
+}
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/wwwroot/ext/custom-plugin2.js:10]:
```suggestion
};
```

File: test/WebSites/CustomUIConfig/Startup.cs [hunk 21]
```diff
@@ -73,14 +73,18 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
                 c.ScriptPresetsPath = "/ext/custom-javascript.js";
-                c.InjectStylesheet("/ext/custom-stylesheet.css");
                 c.InjectJavascript("/ext/custom-javascript.js");
+                c.InjectJavascript("/ext/custom-plugin1.js");
+                c.InjectJavascript("/ext/custom-plugin2.js");
                 c.UseRequestInterceptor("(req) => { req.headers['x-my-custom-header'] = 'MyCustomValue'; return req; }");
                 c.UseResponseInterceptor("(res) => { console.log('Custom interceptor intercepted response from:', res.url); return res; }");
                 c.EnablePersistAuthorization();
 
+                c.ConfigObject.Plugins = ["customPlugin1","customPlugin2"];
```
Comment (martincostello) [inline test/WebSites/CustomUIConfig/Startup.cs:86]:
```suggestion
                c.ConfigObject.Plugins = ["customPlugin1", "customPlugin2"];
```

File: test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs [hunk 22]
```diff
@@ -74,6 +75,29 @@ public async Task SwaggerUIMiddleware_ReturnsInitializerScript(
...
+        [Fact]
+        public async Task IndexUrl_DefinesPlugins()
+        {
+            var client = new TestSite(typeof(CustomUIConfig.Startup)).BuildClient();
+
+            var jsResponse = await client.GetAsync("/swagger/index.js");
+            Assert.Equal(HttpStatusCode.OK, jsResponse.StatusCode);
+
+            var jsContent = await jsResponse.Content.ReadAsStringAsync();
+            Assert.Contains($"\"plugins\":[\"customPlugin1\",\"customPlugin2\"]", jsContent);
```
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.IntegrationTests/SwaggerUIIntegrationTests.cs:87]:
```suggestion
            Assert.Contains("\"plugins\":[\"customPlugin1\",\"customPlugin2\"]", jsContent);
```

**Review by martincostello — APPROVED** [review]

**Comment by jvmlet** [comment]
This is what swagger does ... Not in the app's control.
@martincostello , when do you plan to release next version ?

**Comment by martincostello** [comment]
https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/3060#issuecomment-2355413865

**Comment by martincostello** [comment]
Can someone confirm whether or not the double execution is an issue or not? I don't want to publish 6.8.0 if it turns out there's a bug with this we need to fix.

**Comment by jvmlet** [comment]
This is not double execution , this double instantiation. 
This is what `swagger-ui` does, regardless of the source definition of the plugin (`swashbuckle` or  end-user manual code writing ....)
